### PR TITLE
[DEVX-110] fix: allow shallow copying for Asset and Timeseries

### PR DESCRIPTION
**Pull request by polomani**
https://cognitedata.atlassian.net/browse/DEVX-110

This is the best I can do here to help destructing them easier

So basically those tests show the way Asset and Timeseries behave while shallow copied.
I am still not sure I did the right thing here.

I mean it just makes our code messier, while for the user it's just a matter of calling `assets.map(asset=> asset.toJSON())`
or do something like `assets.map({id, name} => ({id, name}))`

So idk if we should be responsible for people destructing our classes in a poor way... 

Neither `Object.assign` or `spread` operator will ever be able to copy our custom `toJSON`, everything that goes after destruction is not under our control 

I don't like this code at all, btw

### CODE REVIEW

File: src/__tests__/resources/assetList.integration.specs.ts [hunk 1]
```diff
@@ -41,10 +41,4 @@ describe('assetList', () => {
     await client.assets.delete(createdAssets.map(asset => ({ id: asset.id })));
     await client.events.delete(createdEvents.map(event => ({ id: event.id })));
   });
-
-  test('json stringify', async () => {
```
Comment (ghost) [inline src/__tests__/resources/assetList.integration.specs.ts:45]: Why remove?

File: src/__tests__/resources/assets.unit.spec.ts [hunk 2]
```diff
@@ -198,4 +199,49 @@ describe('Asset unit test', () => {
...
+    const items = [
+      {
+        id: 1,
+      },
+      {
+        id: 2,
+      },
+    ];
+
+    test('assets.list().autoPagingToArray()', async () => {
```
Comment (ghost) [inline src/__tests__/resources/assets.unit.spec.ts]: Not meaningful test description

File: src/__tests__/resources/assets.unit.spec.ts [hunk 3]
```diff
@@ -198,4 +199,49 @@ describe('Asset unit test', () => {
...
+      nock(mockBaseUrl)
+        .post(new RegExp('/assets/list'), {})
+        .once()
+        .reply(200, { items });
+
+      const assets = await client.assets.list().autoPagingToArray({ limit: 2 });
+      expect(JSON.stringify(assets)).toBeTruthy();
```
Comment (ghost) [inline src/__tests__/resources/assets.unit.spec.ts]: `toBeTruthy` will I consider to be a weak test here. What do you test with that?

File: src/__tests__/resources/assets.unit.spec.ts [hunk 4]
```diff
@@ -198,4 +199,49 @@ describe('Asset unit test', () => {
...
+      const shallowCopyItems: TypeAsset[] = [...assets, assets[0]];
```
Comment (ghost) [inline src/__tests__/resources/assets.unit.spec.ts]: This is not spreading any objects? Just spreading an array which already works fine...

File: src/__tests__/resources/assets.unit.spec.ts [hunk 5]
```diff
@@ -198,4 +199,49 @@ describe('Asset unit test', () => {
...
+      expect(JSON.stringify(shallowCopyItems)).toBeTruthy();
+      const destructedAsset: TypeAsset = { ...assets[1] };
+      expect(destructedAsset.id).toBeDefined();
+      expect(() => JSON.stringify(destructedAsset)).toThrowError();
+      const destructedJsonAsset: TypeAsset = { ...assets[1].toJSON() };
+      expect(JSON.stringify(destructedJsonAsset)).toBeTruthy();
+    });
+
+    test('assets.list()', async () => {
```
Comment (ghost) [inline src/__tests__/resources/assets.unit.spec.ts]: I don't see the need for this test? Duplicate of the one above? If we are using autoPaginate or not shouldn't matter or no?

File: src/__tests__/resources/timeSeriesList.unit.spec.ts [hunk 6]
```diff
@@ -99,4 +101,38 @@ describe('TimeSeriesList class unit test', async () => {
...
+    const items = [
+      {
+        id: 1,
+      },
+      {
+        id: 2,
+      },
+    ];
+
+    test('timeseries.list().autoPagingToArray()', async () => {
```
Comment (ghost) [inline src/__tests__/resources/timeSeriesList.unit.spec.ts]: Bad test description. If this test fails what was the cause of the problem by looking at the logs?

File: src/resources/classes/timeSeries.ts [hunk 7]
```diff
@@ -1,53 +1,44 @@
...
+  public description: string;
+  public securityCategories?: number[];
+  public createdTime: Date;
+  public lastUpdatedTime: Date;
+  public id: CogniteInternalId;
 
   constructor(client: CogniteClient, props: GetTimeSeriesMetadataDTO) {
-    super(client, props);
+    super(client);
+    this.externalId = props.externalId;
```
Comment (ghost) [inline src/resources/classes/timeSeries.ts:30]: We could potentially add support in `BaseResource` for a list of prop names like `['externalId', 'name', 'isString', ...]`. And then use that list to create the `toJSON` return value and we could send the props to `BaseResource` as before.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/resources/classes/timeSeries.ts (hunk 7 above)
Comment (polomani) [reply src/resources/classes/timeSeries.ts:30]: and the force casting to the type we need? 'cause we will lose the type information
also how will you able to remember to update that list of strings every time you update the interface?

File: src/__tests__/resources/assets.unit.spec.ts (hunk 5 above)
Comment (polomani) [reply src/__tests__/resources/assets.unit.spec.ts]: the one above returns `Asset[]`, but this one returns `AssetList`
bug?

File: src/__tests__/resources/assets.unit.spec.ts (hunk 4 above)
Comment (polomani) [reply src/__tests__/resources/assets.unit.spec.ts]: you are right, will remove it

File: src/__tests__/resources/assets.unit.spec.ts (hunk 3 above)
Comment (polomani) [reply src/__tests__/resources/assets.unit.spec.ts]: basically just that it's not failing, maybe snapshot instead?

File: src/__tests__/resources/assetList.integration.specs.ts (hunk 1 above)
Comment (polomani) [reply src/__tests__/resources/assetList.integration.specs.ts:45]: moved to unit test intead

File: src/__tests__/resources/assets.unit.spec.ts [hunk 8]
```diff
@@ -198,4 +199,44 @@ describe('Asset unit test', () => {
...
+      {
+        id: 2,
+      },
+    ];
+
+    beforeEach(() => {
+      nock.cleanAll();
+      nock(mockBaseUrl)
+        .post(new RegExp('/assets/list'))
+        .thrice()
```
Comment (ghost) [inline src/__tests__/resources/assets.unit.spec.ts:217]: Why `thrice`? This runs before each and not before all

File: src/__tests__/resources/assets.unit.spec.ts [hunk 9]
```diff
@@ -198,4 +199,44 @@ describe('Asset unit test', () => {
...
+      expect(() => JSON.stringify(assets[0])).not.toThrow();
+    });
+
+    test('change context for asset utility methods', async () => {
+      const assets = await client.assets.list().autoPagingToArray();
+      const utilMethod = assets[0].children;
+      const result = await utilMethod();
+      const resultAfterBind = await utilMethod.call(null);
+      expect({ ...result[0] }).toEqual({ ...resultAfterBind[0] });
+      expect([{ ...result[0] }, { ...result[1] }]).toEqual(items);
```
Comment (ghost) [inline src/__tests__/resources/assets.unit.spec.ts:233]: Excellent!

**Review by ghost — APPROVED** [review]

File: src/__tests__/resources/assets.unit.spec.ts (hunk 8 above)
Comment (polomani) [reply src/__tests__/resources/assets.unit.spec.ts:217]: one of tests requires 3 calls, I didn't want to have a special case for that..
